### PR TITLE
WIP txHandler: do not rebroadcast to peers sent duplicate messages

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

When the transaction deduplication was added it had a side effect of re-sending transactions to peers sent a message the handler has seen before. This PR fixes this.

Before dedup if A sent T, all but A get T, then, if B sent us T, noone gets T again
With dedup, if A sent T, all but A get T, then, if B sent us T, noone gets T again.
With this PR: if A and B sent T almost simultaneously all but A and B get T

### Implementation overview:

1. `txSaltedCache` how stores map of seen peers for a particular message
2. `CheckAndPut` returns this map to to get rid additional lookups.
3. The map is later used in `RelayArray` to let networking know whom to skip.

### Implementation Details

`CheckAndPut` is not more complex because it needs to update the map value even if there is a match.
The original idea with fast check under a read lock is preserved but `innerCheck` also returns a current value and a "page" (cur/prev map) where it was found in order to update. Note `innerCheck` can return prev page and this also need to be considered when running an update with a write lock taken.
The `found && senderFound` denotes a new fast path without modification underlying maps. 

Note, a reference data struct (`*sync.Map` in this case) is crucial to have the implementation to work because of the following scenario:
1. Received txn A from N1, wrote to the cache, the the cache value (peers list) is attached to a work item `txBacklogMsg` (`wi `) of this transaction.
2. Received txn A from N2, updated the peers list. The reference in `wi` is updated automatically.
3. It is time to relay, give the network the same peers list
4. Received txn A from N3, updated to the cache value (peers list)
5. The network has updated peers list with N3
6. After broadcasting received txn A from N4, updated to the cache value (peers list), but b/c it is a duplicate it would not make its path to the network.

## Test Plan

Added a new unit test

## Benchmarks

master
```
BenchmarkDigestCaches/data.digestCacheMaker/threads=1-8         	 1000000	      1575 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=1-8         	  769404	      1886 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=4-8         	 1347638	       992.8 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=4-8         	  777824	      2042 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=16-8        	 1248975	       984.3 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=16-8        	  709402	      2234 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=128-8       	 1233063	       996.1 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=128-8       	  619723	      2289 ns/op
```

feature
```
BenchmarkDigestCaches/data.digestCacheMaker/threads=1-8         	  978464	      1460 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=1-8         	  575578	      2303 ns/op
BenchmarkDigestCaches/data.saltedCacheDupMaker/threads=1-8      	  550164	      2238 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=4-8         	 1395024	       877.9 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=4-8         	  499202	      2773 ns/op
BenchmarkDigestCaches/data.saltedCacheDupMaker/threads=4-8      	  651159	      1883 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=16-8        	 1308243	       948.8 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=16-8        	  430381	      3130 ns/op
BenchmarkDigestCaches/data.saltedCacheDupMaker/threads=16-8     	  954093	      1338 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=128-8       	 1305469	       952.8 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=128-8       	  443257	      3026 ns/op
BenchmarkDigestCaches/data.saltedCacheDupMaker/threads=128-8    	 1756513	       686.5 ns/op
```

### CODE REVIEW

File: data/txDupCache.go [hunk 1]
```diff
@@ -170,17 +192,17 @@ func (c *txSaltedCache) innerSwap(scheduled bool) {
 
 	if scheduled {
 		// updating by timer, the prev size is a good estimation of a current load => preallocate
-		c.cur = make(map[crypto.Digest]struct{}, len(c.prev))
+		c.cur = make(map[crypto.Digest]cacheValue, len(c.prev))
 	} else {
 		// otherwise start empty
-		c.cur = map[crypto.Digest]struct{}{}
+		c.cur = map[crypto.Digest]cacheValue{}
```
Comment (iansuvak) [inline data/txDupCache.go:198]: The else case here it means it's not scheduled. We should probably pre-allocate maxsize here since it's one of two cases:

1. The node has just started up
2. The cache rotated due to previous cache being full

In case of high traffic we would be adding extra allocation churn here since we would be hitting case 2 often and allocating incrementally. If the network is not at capacity we would free up over-allocated memory after the first scheduled interval

File: data/txDupCache.go [hunk 2]
```diff
@@ -250,13 +308,26 @@ func (c *txSaltedCache) CheckAndPut(msg []byte) (*crypto.Digest, bool) {
...
+	delete(c.prev, *d)
+}
+
+// Len returns size of a cache
+func (c *txSaltedCache) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return len(c.cur) + len(c.prev)
 }
```
Comment (AlgoAxel) [inline data/txDupCache.go:331]: could len be managed proactively as two atomic ints? If we did that, we could avoid taking a lock on len checks I think?

File: data/txDupCache_test.go [hunk 3]
```diff
@@ -371,3 +371,132 @@ func benchmarkDigestCache(b *testing.B, m cacheMaker, numThreads int) {
...
+	type snd struct {
+		id int
+	}
+
+	d, v, p, found := cache.innerCheck([]byte{1})
+	require.False(t, found)
+	require.Nil(t, p)
+	require.Nil(t, v)
+	require.NotNil(t, d)
+	require.NotEmpty(t, d)
```
Comment (AlgoAxel) [inline data/txDupCache_test.go:394]: This is just to confirm the data structure is empty before the test starts, or is there other utility? A brief comment might help. If you're just looking to confirm it's empty, why not use your Len function?

File: data/txDupCache.go [hunk 4]
```diff
@@ -191,51 +213,87 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
...
+	// copy vals since its owned by the cache and may be concurrently modified
+	copy := func(in cacheValue) cacheValue {
+		out := make(cacheValue, len(in))
+		for p := range in {
+			out[p] = struct{}{}
+		}
+		return out
+	}
+
 	c.mu.RLock()
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txDupCache.go:243]: this is a function definition, and both call below are under a lock.

File: data/txHandler.go [hunk 5]
```diff
@@ -564,10 +570,11 @@ func (handler *TxHandler) dedupCanonical(ntx int, unverifiedTxGroup []transactio
 func (handler *TxHandler) processIncomingTxn(rawmsg network.IncomingMessage) network.OutgoingMessage {
 	var msgKey *crypto.Digest
 	var isDup bool
+	var peers map[network.Peer]struct{}
 	if handler.msgCache != nil {
 		// check for duplicate messages
 		// this helps against relaying duplicates
-		if msgKey, isDup = handler.msgCache.CheckAndPut(rawmsg.Data); isDup {
+		if msgKey, peers, isDup = handler.msgCache.CheckAndPut(rawmsg.Data, rawmsg.Sender); isDup {
```
Comment (iansuvak) [inline data/txHandler.go:577]: I'm probably missing something but I'm not sure how this is meant to work. 

- when first copy of the transaction T comes through `peers` here will be a copy of the peers value from CheckAndPut and peers set will have the single peer A. This copy of peers gets added onto the queue
- second copy of transaction T comes from peer B . It will create a copy of the existing map and add B to it  and update the value in the cache correctly.
- when we Remember T from the message put on the backlog by A the `peers` value we will Relay it but `wi.peers` will still only contain peer A since it had it's copy from the first pass.

Comment (algorandskiy) [reply data/txHandler.go:577]: oh, you are absolutely right. It worked before I introduced map copies since they were accessed in wsnetwork latter and caused raises. Good catch, working on a test and will fix this.

File: data/txDupCache.go (hunk 1 above)
Comment (algorandskiy) [reply data/txDupCache.go:198]: but without load it would cause memory waste. Maybe start with 1/3 of the max size or something?

File: data/txHandler.go (hunk 5 above)
Comment (algorandskiy) [reply data/txHandler.go:577]: Switched to *sync.Map since it appears to be an ideal use

File: data/txDupCache.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txDupCache.go]: no copying anymore - it ended up to be incorrect.

File: data/txDupCache.go (hunk 1 above)
Comment (iansuvak) [reply data/txDupCache.go]: But for only a maximum of 1 minute or whatever the timeout is set to be. If we are rotating because of capacity it's likely that the next one will fill up as well and we will have to allocate more memory to get it back up to maximum. 

What's odd to me is that if we rotated on a schedule when the cache was 99% full we will start with 99% of maxsize pre-allocated but if we actually fill it up 100% we will start with it empty.

If we never fill up we will never-allocate full maxsize other than at node startup. 

A compromise would be to start at 1/3 maxsize when starting the node up (or empty) and then re-allocate max if we've expired but it's a bit of a complication for marginal benefits

Comment (algorandskiy) [reply data/txDupCache.go]: use len(c.prev) unconditionally now

File: data/txHandler_test.go [hunk 6]
```diff
@@ -1097,6 +1099,80 @@ loop:
...
+	// wait until txn propagates in handler
+	var timeout time.Duration
+	waitTime := 10 * time.Millisecond
+	for tp.PendingCount() < 1 {
+		time.Sleep(waitTime)
+		timeout += waitTime
+		if timeout >= time.Second {
+			require.Fail(t, "timed out waiting for txn to propagate")
+		}
+	}
```
Comment (AlgoAxel) [inline data/txHandler_test.go:1186]: You can use "Eventually" to block on a condition for up to a maximum time, failing if it exceeds the time:

https://pkg.go.dev/github.com/stretchr/testify/require#Eventually

File: data/txHandler_test.go [hunk 7]
```diff
@@ -1097,6 +1099,80 @@ loop:
...
 
+type dupMockNetwork struct {
+	mocks.MockNetwork
+	relay []map[network.Peer]struct{}
+}
+
+func (network *dupMockNetwork) RelayArray(ctx context.Context, tag []protocol.Tag, data [][]byte, wait bool, except map[network.Peer]struct{}) error {
+	network.relay = append(network.relay, except)
+	return nil
+}
```
Comment (AlgoAxel) [inline data/txHandler_test.go]: Cool Spy Mocks :)

**Review by AlgoAxel — DISMISSED** [review]
One opportunity to use `Eventually`, but up to you if you take it.

File: data/txDupCache.go [hunk 8]
```diff
@@ -25,25 +25,33 @@ import (
...
 	cur  map[crypto.Digest]struct{}
 	prev map[crypto.Digest]struct{}
 
 	maxSize int
 	mu      deadlock.RWMutex
 }
 
+// digestCache is a rotating cache of size N accepting crypto.Digest as a key
+// and keeping up to 2*N elements in memory
+type digestCache struct {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txDupCache.go:45]: originally `digestCache` was a base for `saltedCache`, but now they diverged in what is a cache value. I tried to somehow have reused code but it did not work out so there `digestCacheBase` and `digestCacheData` with different map values but the same idea. Nothing prevents from removing these aux types

[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txDupCache.go]: done

**Review by iansuvak — DISMISSED** [review]

**Review by AlgoAxel — DISMISSED** [review]
Since my last approval, the change made is to remove the cache base structure abstraction and simply use a `txSaltedCache`. those changes also look good, so reapproving.

File: data/txDupCache.go [hunk 9]
```diff
@@ -191,51 +192,74 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
...
+				return d, vals, true
+			}
+			page = &c.cur
 		}
 	}
 
+	// at this point we know that either:
+	// 1. the message is not in the cache
+	// 2. the message is in the cache but from other senders
+	if found && !senderFound {
```
Comment (algonautshant) [inline data/txDupCache.go:257]: According to the comment above, this should be sufficient:
```suggestion
	if found {
```

File: data/txDupCache.go [hunk 10]
```diff
@@ -191,51 +192,74 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
...
+		vals.Store(sender, struct{}{})
+		(*page)[*d] = vals
+		return d, vals, true
+	}
+
 	if len(c.cur) >= c.maxSize {
```
Comment (algonautshant) [inline data/txDupCache.go:263]: Before, the size was checked at every new addition to the cache.
Now, it is only checked for message addition, and not sender addition. Is this okay?

File: data/txDupCache.go [hunk 11]
```diff
@@ -191,51 +192,74 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
...
+		// there is match with prev page, update the value with data possible added in between locks
+		vals, found = c.prev[*d]
+	} else { // not found or found in cur page
 		// Do another check to see if another copy of the transaction won the race to write it to the cache
-		// Only check current to save a lookup since swaps are rare and no need to re-hash
-		if _, found := c.cur[*d]; found {
-			return d, found
+		// Only check current to save a lookup since swap is handled in the first branch
+		vals, found = c.cur[*d]
+		if found {
```
Comment (algonautshant) [inline data/txDupCache.go:246]: A lot of comments here, but still very hard to parse.

File: data/txDupCache.go [hunk 12]
```diff
@@ -191,51 +192,74 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
...
+	// keep lock - it is needed for copying vals in defer
+	senderFound := false
 	if found {
-		return d, found
+		if _, senderFound = vals.Load(sender); senderFound {
+			c.mu.RUnlock()
+			return d, vals, true
+		}
 	}
+	c.mu.RUnlock()
```
Comment (algonautshant) [inline data/txDupCache.go:225]: Do you need a write lock for all the cases below?
If only the sender is missing, it can be added without a write lock right?

File: data/txDupCache.go (hunk 10 above)
Comment (algorandskiy) [reply data/txDupCache.go:263]: I guess so since we are counting incoming txns that is our primary target.

File: data/txHandler_test.go (hunk 6 above)
Comment (algorandskiy) [reply data/txHandler_test.go]: done

**Review by algonautshant — DISMISSED** [review]
Looks good. 

Many more locks, map lookups, code complexity and memory footprint added for possibly very limited gain. 
Will be nice to evaluate this tradeoff thoroughly and be prepare to reverse this if it is not  sufficiently helpful.

File: data/txDupCache.go [hunk 13]
```diff
@@ -191,53 +192,76 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
...
+	} else { // not found or found in cur page
 		// Do another check to see if another copy of the transaction won the race to write it to the cache
-		// Only check current to save a lookup since swaps are rare and no need to re-hash
-		if _, found := c.cur[*d]; found {
-			return d, found
+		// Only check current to save a lookup since swap is handled in the first branch
+		vals, found = c.cur[*d]
+		if found {
+			if _, senderFound = vals.Load(sender); senderFound {
+				return d, vals, true
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txDupCache.go:248]: it is hard to test since the value need to appear in a current page between read write locks

**Review by iansuvak — DISMISSED** [review]

**Comment by algorandskiy** [comment]
A current delay between accepting and and relaying is:
1. Hashing + cache access
2. Decoding
3. Signature verification within 2ms [batch](https://github.com/algorand/go-algorand/pull/4621/files#diff-d3741bc12fa9faf0a148f027517dbfde4942668add6ee4ac7b4032543db4b0f5R72)
4. txpool Check + Remember delay

I.e. there is some delay but might not be enough to collect enough duplicates to make re-broadcasting filtering effective.
TODO, as a follow up PR: benchmark and and implement ~50ms (?) delay before re-broadcasting in order to collect as much duplicates as possible without introducing too much tx latency.

**Comment by algorandskiy** [comment]
Do not merge - the benchmark showed 6.9k tps vs 7.7k tps on master, so networking code appears to be much slower now.

File: data/txDupCache.go (hunk 12 above)
Comment (algorandskiy) [reply data/txDupCache.go:225]: this is valid, I'll refactor the code. Thank you. I did not update locking after switching from map to sync.Map

File: data/txDupCache_test.go [hunk 14]
```diff
@@ -315,7 +315,7 @@ func (p *digestCachePusher) push() {
 func (p *saltedCachePusher) push() {
 	var d [crypto.DigestSize]byte
 	crypto.RandBytes(d[:])
-	p.c.CheckAndPut(d[:]) // saltedCache hashes inside
+	p.c.CheckAndPut(d[:], struct{}{}) // saltedCache hashes inside
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txDupCache_test.go:330]: yes, updated the benchmark and posted results

File: data/txDupCache.go (hunk 12 above)
Comment (algorandskiy) [reply data/txDupCache.go]: pushed an updated

File: data/txDupCache.go [hunk 15]
```diff
@@ -191,53 +192,76 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
...
-	d, found := c.innerCheck(msg)
+	d, vals, page, found := c.innerCheck(msg)
 	salt := c.curSalt
-	c.mu.RUnlock()
 	// fast read-only path: assuming most messages are duplicates, hash msg and check cache
+	// keep lock - it is needed for copying vals in defer
+	senderFound := false
 	if found {
-		return d, found
+		if _, senderFound = vals.Load(sender); senderFound {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txDupCache.go]:
```
BenchmarkDigestCaches/data.digestCacheMaker/threads=1-8         	  944719	      1435 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=1-8         	  585506	      2342 ns/op
BenchmarkDigestCaches/data.saltedCacheDupMaker/threads=1-8      	  568954	      2254 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=4-8         	 1327412	       893.9 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=4-8         	  495157	      2733 ns/op
BenchmarkDigestCaches/data.saltedCacheDupMaker/threads=4-8      	  627554	      2041 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=16-8        	 1287717	      1023 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=16-8        	  382932	      3128 ns/op
BenchmarkDigestCaches/data.saltedCacheDupMaker/threads=16-8     	  981458	      1280 ns/op
BenchmarkDigestCaches/data.digestCacheMaker/threads=128-8       	 1307422	       988.9 ns/op
BenchmarkDigestCaches/data.saltedCacheMaker/threads=128-8       	  406440	      3055 ns/op
BenchmarkDigestCaches/data.saltedCacheDupMaker/threads=128-8    	 1475484	       787.5 ns/op
```
idk, kind of the same. 

>The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. 

It appears to be a sharded map hashmap, and according to (1) it is better to read rather then rewrite

[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txDupCache.go]: `sync.Map.Store` actually calls `Load` [at the very beginning ](https://github.com/golang/go/blob/dev.boringcrypto.go1.17/src/sync/map.go#L137-L141) so in makes sense to call Store directly.
But anyway, allocation new sync.Map per new txn appears to be a main contributor to the TPS slowdown.
